### ISSUE TITLE
Error after reducing number of layers and input size

### ISSUE THREAD

**Issue by Addhi86** [vlfeat/matconvnet#850]
I have recently started implementing Binary classification using CNN.  I have prepared the code for the Binary Classification (aircraft (6056 images)and non-aircraft(5760)).

```
net.meta.inputSize = [28 28 3] ;
net.meta.trainOpts.learningRate = 0.00001; 
net.meta.trainOpts.numEpochs = 50 ;
% net.meta.trainOpts.weightDecay = 0.01 ;
net.meta.trainOpts.batchSize = 60 ;
opts.train.gpus = [];
f=1/100 ;
net.layers = {} ;
net.layers{end+1} = struct('type', 'conv','weights', {{f*randn(5,5,3,50, 'single'), zeros(1, 50, 'single')}},'stride', 1,'pad', 0);
net.layers{end+1} = struct('type', 'pool','method', 'max', 'pool', [2 2], 'stride', 2, 'pad', 0) ;
net.layers{end+1} = struct('type', 'conv', 'weights', {{f*randn(5,5,50,30, 'single'),zeros(1,30,'single')}}, 'stride', 1, 'pad', 0) ;
net.layers{end+1} = struct('type', 'pool', 'method', 'max', 'pool', [2 2], 'stride', 2, 'pad', 0) ;
net.layers{end+1} = struct('type', 'conv', 'weights', {{f*randn(4,4,30,50, 'single'),  zeros(1,50,'single')}}, 'stride', 1, 'pad', 0) ;
net.layers{end+1} = struct('type', 'relu') ;
net.layers{end+1} = struct('type', 'conv', 'weights', {{f*randn(1,1,50,2, 'single'), zeros(1,2,'single')}}, 'stride', 1, 'pad', 0) ;
net.layers{end+1}= struct('type', 'dropout', 'rate', 0.15);
net.layers{end+1} = struct('type', 'softmaxloss') ;
 net.meta.classes.name =arrayfun(@(x)sprintf('%d',x),0:1,'UniformOutput',false);
[net, info] = cnn_train( net, imdb, @getBatch, net.meta.trainOpts, opts.train, 'val', find( imdb.images.set == 2 ) ) ;
```

After that, I am following the architecture of MINST example.  if I make the input image size `50x50`  or reduce the number of layers. I am getting this error. What are the possible reasons for this error?
![image](https://cloud.githubusercontent.com/assets/22164098/22172545/0165fbe0-dffe-11e6-970d-017162fe2882.png)

**Comment by H-Seyf** [comment]
Hello,
I had faced the same problem. this might help:
I also faced a new problem, h
![untitled](https://cloud.githubusercontent.com/assets/20422644/22244759/1cd26058-e242-11e6-8f56-7e4e41346999.jpg)
ope to get help for solving it ... 
I think the error refers to computing err and it's not related to reducing input size.
The problem arises when the error_binary function in the cnn_train file wants to compute the err;
the 'predictions' variable is a 3-D vector and Labels is a 1-D vector so dimension obviously don't match for applying inner product....
soultion:
1. go to cnn_train and change opts.errorFunction = 'multicalss'  to  opts.errorFunction = 'binary' 
2. go to cnn_train and do these modification on error_binary function

function err = error_binary(opts, labels, res)
predictions = gather(res(end).x) ;
error = bsxfun(@times, predictions, labels) < 0 ;
err = sum(error(:)) ;
3. make sure you are using {+1,-1} as labels
I have done this modifications and got these results:

![untitled](https://cloud.githubusercontent.com/assets/20422644/22245064/58b9e298-e243-11e6-9cec-e4232f9b1ced.jpg)
![capture2](https://cloud.githubusercontent.com/assets/20422644/22245067/5b2deca4-e243-11e6-8ca6-04655bb52497.JPG)


after train epoch 1, objective values comes to zero.
I use these specifications:
net.meta.inputSize = [224 224 3] ;
net.meta.trainOpts.learningRate = [0.001 0.001] ;
net.meta.trainOpts.numEpochs = 5 ;
net.meta.trainOpts.batchSize = 5 ;

also there are 235 Images for Training and 100 Images for validation
 I think there are some problem with the results...
Any help is appreciated..
Best Regards

**Comment by Addhi86** [comment]
@H-Seyf  @vedaldi  @lenck  

is it the bug in the code for the binary classification to do this change for the error function in the `cnn_train`?
```
predictions = gather(res(end).x) ;
error = bsxfun(@times, predictions, labels) < 0 ;
err = sum(error(:));
```
When I have changed like that
This is my network architecture
![image](https://cloud.githubusercontent.com/assets/22164098/22356427/6339804a-e485-11e6-97d7-d83a3d6ad127.png)

the training output is 
![image](https://cloud.githubusercontent.com/assets/22164098/22356443/81b959a0-e485-11e6-87a1-a3415540b27d.png)

While testing like this

![image](https://cloud.githubusercontent.com/assets/22164098/22356459/96dc1336-e485-11e6-8c08-c198885b6da2.png)

The variable `Predic_class` has always '1'. for all the positive and negative image test.

I want to ask couple of thing if any one know..

q1: What is the difference between `Binerr` and `Objective` in the above training figure.
q2: `Binerr` should also converge to zero like `Objective` ?

**Comment by H-Seyf** [comment]
@Addhi86 
It was an offer for modification of error function. It may not be optimal but seems working.
Assuming you are using {-1,1} as labels, change this line and please let me know if it works:
[bestScore(i,:), best(i,:)]= max(abs(scores));
It seems that object value should converge to zero but should not reach to it. In my binary training as I mentioned, after epoch 1 it reaches to zero. please let me know about your binary training procedure.
I used MNIST demo. 
batch size is set to 5;
learning rate is set to [0.001 0.001];
@fengyunxiaozi
fengyyunxiaozi, you have told that when object value reaches to zero it means the network is overtrained. I have two questions:
1. what can we do for solving these problem? changing batch-size and learning-rate might help?
2. Is it possible for network to be over-trained after just one epoch?
any help is appreciated
@titubeta

**Comment by Addhi86** [comment]
@H-Seyf 
Binary classification is working now.. i have set positive images with label 2 and negative image label 1. 
The `imdb` variable is llike that:
```
imdb.images.data(:,:,:,i)=im2single(data{i,:});
imdb.images.label=label';
imdb.images.set=sets';
imdb.meta.sets = {'train', 'val', 'test'} ;
imdb.meta.classes ={'notplane','plane'};
```
Can you please tell me about the  difference between Binerr and Objective in the above training figure.

**Comment by H-Seyf** [comment]
@Addhi86 
Hello, Sorry for late respone,
I just like you don't know the exact definition of Objective and I'm waiting for some how "precise" answer  to this question. but it might help:
These variables are computed in cnn_train.m, so analyzing this file can be useful.

% -------------------------------------------------------------------------
function err = error_binary(opts, labels, res)
predictions = gather(res(end-1).x) ;
error = bsxfun(@times, predictions, labels) < 0 ;
err = sum(error(:)) ;
% -------------------------------------------------------------------------

here the "error" is computed. 

% -------------------------------------------------------------------------
stats.objective = errors(1) ;
% -------------------------------------------------------------------------
 
but as we see the "errors" is put in Objective, so Objective is some how sum of the errors.
*****
You mentioned that have successfully done Binary classification. Can you please share the code and specification of parameters?
Best Regards
